src/backend/distributed/commands/extension.c

```diff
@@ -491,6 +495,139 @@ PreprocessAlterExtensionUpdateStmt(Node *node, const char *queryString)
+	 */
+	MarkExistingObjectDependenciesDistributedIfSupported();
+}
+
+
+/*
+ * MarkAllExistingObjectsDistributed marks all objects that could be distributed by
+ * resolving dependencies of "existing distributed tables" and "already distributed
+ * objects" to introduce the objects created in older versions of Citus to distributed
+ * object infrastructure as well.
+ *
+ * Note that this function is not responsible for ensuring if dependencies exist on
+ * nodes and satisfying these dependendencies if not exists, which is already done by
+ * EnsureDependenciesExistOnAllNodes on demand. Hence, this function is just designed
+ * to be used when "ALTER EXTENSION citus UPDATE" is executed.
+ * This is because we want to add existing objects that would have already been in
+ * pg_dist_object if we had created them in new version of Citus to pg_dist_object.
+ */
+static void
+MarkExistingObjectDependenciesDistributedIfSupported()
+{
+	ListCell *listCell = NULL;
+
+	/* resulting object addresses to be marked as distributed */
+	List *resultingObjectAddresses = NIL;
+
+	/* resolve dependencies of distributed tables */
+	List *distributedTableOidList = DistTableOidList();
+
+	foreach(listCell, distributedTableOidList)
+	{
+		Oid distributedTableOid = lfirst_oid(listCell);
+
+		ObjectAddress tableAddress = { 0 };
+		ObjectAddressSet(tableAddress, RelationRelationId, distributedTableOid);
+
+		List *distributableDependencyObjectAddresses =
+			GetDistributableDependenciesForObject(&tableAddress);
+
+		resultingObjectAddresses = list_concat(resultingObjectAddresses,
+											   distributableDependencyObjectAddresses);
+	}
+
+	/* resolve dependencies of the objects in pg_dist_object*/
+	List *distributedObjectAddressList = GetDistributedObjectAddressList();
+
+	foreach(listCell, distributedObjectAddressList)
+	{
+		ObjectAddress *distributedObjectAddress = (ObjectAddress *) lfirst(listCell);
+
+		List *distributableDependencyObjectAddresses =
+			GetDistributableDependenciesForObject(distributedObjectAddress);
+
+		resultingObjectAddresses = list_concat(resultingObjectAddresses,
+											   distributableDependencyObjectAddresses);
+	}
+
+	/* reorder object addresses list according to dependency order */
+	List *sortedObjectAddresses = OrderObjectAddressListInDependencyOrder(
+		resultingObjectAddresses);
+
+	/* subtract the ones that are already in pg_dist_object */
+	sortedObjectAddresses = list_difference(sortedObjectAddresses,
+											distributedObjectAddressList);
+
+	/*
+	 * Make sure that no new nodes are added after this point until the end of the
+	 * transaction by taking a RowShareLock on pg_dist_node, which conflicts with the
+	 * ExclusiveLock taken by master_add_node.
+	 * This guarantees that newly added nodes will have the object, because they will
+	 * get it in master_add_node after this transaction finishes and the pg_dist_object
+	 * record becomes visible.
+	 */
+	LockRelationOid(DistNodeRelationId(), RowShareLock);
```

Comment:
I don't think this line is particularly useful in this context.

In practice we don't really have to care about concurrency because if we're doing an upgrade it means all concurrent Citus operations error out until this transaction finishes.

If we did care, this lock would not do much, since this function does not affect whether or not a node has an object.

Reply:
makes sense, removing the lock stmt